zerver/tornado/event_queue.py

```diff
@@ -522,6 +524,30 @@ def fetch_events(query: Mapping[str, Any]) -> Dict[str, Any]:
                 raise BadEventQueueIdError(queue_id)
             if user_profile_id != client.user_profile_id:
                 raise JsonableError(_("You are not authorized to get events from this queue"))
+            if last_event_id < client.event_queue.newest_pruned_id:
+                # This code path is intended to catch misbehaving
+                # browsers that have incorrectly restored/reloaded a
+                # Zulip app browser window from cache, rather than
+                # re-fetching /, resulting in a window initialized
+                # from a stale version of page_params.  If this
+                # happens less than queue_lifespan_secs after the last
+                # request from the original browser window, the event
+                # queue will still exist, and requests from the new
+                # window will reach it.
+                #
+                # We detect this situation by looking at whether
+                # `last_event_id` (the ID the client has acked all
+                # events through, and is requesting all events newer
+                # than) is lower than the ID any events that have
+                # already been pruned.  If so, the request is invalid,
+                # and we return a special error for this situation.
+                #
+                # The client is responsible for doing something
+                # appropriate (usually destroying its event queue and
+                # reloading), because a harmless version of this sort
+                # of ID inversion is possible for cancelled/stale
+                # requests being retried at the network layer.
+                raise RequestedPrunedEventsError(last_event_id)
```

Comment:
Will we see this in logs in case we want to investigate further?

Reply:
We'd see a 400 error response.  Our default logging policy does not include details like the `code`.  We could probably add something here:
```
    # Log some additional data whenever we return certain 40x errors                                 
    if 400 <= status_code < 500 and status_code not in [401, 404, 405]:                              
```

It's not hard to reproduce this code path + reload by patching `do_events_register` to be:
```
    if len(events) > 0:                                                                              
        ret['last_event_id'] = events[-1]['id']                                                      
    else:                                                                                            
        ret['last_event_id'] = -10
```

if you want to play around with this